src/rgw/driver/d4n/d4n_directory.cc

```diff
@@ -85,11 +85,17 @@ int ObjectDirectory::exist_key(const DoutPrefixProvider* dpp, CacheObj* object,
   return std::get<0>(resp).value();
 }
 
-int ObjectDirectory::set(const DoutPrefixProvider* dpp, CacheObj* object, optional_yield y) 
+int ObjectDirectory::set(const DoutPrefixProvider* dpp, CacheObj* object, bool nx, optional_yield y) 
 {
-  std::string key = build_index(object);
+  if (nx) {
```

Comment:
instead of checking if the key already exists, why not just set whatever value is passed in. This will reduce one call to the redis server to check for if the key exists. In the filter driver also, instead of checking if set returns an error an then updating the required fields, all the required fields can be filled in and the set method can be invoked. 
How about fields that aren't filled in? Will the old values be retained? 

Even in `get`, can we eliminate the check for key existence? `get` can just return an error if the value does not exist

Will the above work?

Reply:
There are some areas in the read workflow that require checking for the block in case it exists so there is no overwriting of existing metadata, so for those situations, I have added the `nx` bool. If this bool is set to `true`, that means that the values should only be set if the block does not exist. If it is set to `false`, that means that the values can be set regardless of whether the block exists or not.

Reply:
ok, so the existing values do get overwritten - so this can be done in two ways - either you add a check as you have already done, or you enforce that a get needs to be called before set in case of an existing key (by documenting this in set). Also in the workflows where values need to be modified, please ensure that extra `update_field` calls are avoided and only one `set` is used to modify all values

Reply:
I removed the `exist_key` call and restructured the workflow to match the outline you provided